### PR TITLE
(Neo)Forge support

### PR DESCRIPTION
**Pull request by RubixDev**
This is what I've got so far for adding cross-platform support. See #4.

As you noted in that issue:

> Another problem is about the versionPredicate syntax design. Possible workarounds:
> 
> 1. Keep using the current fabric syntax and perform translation to the target platform. Compatibly with the current design, but might be strange for forge users
> 2. Simply use the syntax that the platform provides. Disadvantage: in multi-platform projects, an annotation instance with version predicate cannot be directly reused in different platforms
> 3. Design a new universal syntax, and translate to the target platform. tho I'll prefer 1. than this
> 4. Shadows let's say fabric's version checking codes into the mod, and use it for all platforms. A bit too tricky for me

I have had similar thoughts. Currently number 2 is implemented, but only because it's the easiest. I don't think this should be the final solution as it will be very annoying for multi-platform users. I think number 3 would be the cleanest, but also the most effort. Number 1 would be nice as it preserves the current syntax, and I don't think it'd be a huge deal for Forge users to have to use that syntax as it's pretty intuitive. As for number 4, that sounds like an option but I'm afraid I have too little experience with Gradle stuff to get that working.

---

A few other notable things about the current changes:

- I changed the Minecraft version to 1.20.2 as that seems to be the first version supported by NeoForge. I kept the java `sourceCompatability` and `targetCompatibilty` as `"1.8"` though in hopes to keep support for old versions on Fabric and Forge.
- I don't know about the Gradle publishing stuff. I think I kept the previous configuration, but it's best if you will take a look.
- The GitHub workflows probably still need to be updated.
- I haven't tested any of this in any way yet.

You are welcome to make changes directly to this PR and maybe implement the version translation as I currently don't have a clear vision of how I would do that.

### CODE REVIEW

**Comment by Fallen-Breath** [comment]
Wow that's a huge amount of changes, thanks in advanced

Some questions:

1. Any instructions on how you set up this project, like basing on an example mod? Providing the source might help for better maintainability on the setup
2. I see gradle stuffs are totally rewritten. It's fine for the maven publish thing, I can fix it latter. Have you test if jitpack works
3. By using jitpack, you can test if your PR works as expected in both platforms, and wide enough mc version range (e.g. 1.14.4 fabric with old loader and java8)

For the predicate syntax thing, idk if it's possible to make an automatic translation, to detect (should be detectable right) translate any of the fabric / forge dialect into the current runtime dialect. Don't need to cover 100% cases should be fine (for those special cases, just please use the naive dialect). This can be done at the future I think

File: build.gradle [hunk 1]
```diff
@@ -1,70 +1,69 @@
...
+		mappings "net.fabricmc:yarn:${project.yarn_mappings}:v2"
 	}
-}
 
-// ensure that the encoding is set to UTF-8, no matter what the system default is
-// this fixes some edge cases with special characters not displaying correctly
-// see http://yodaconditions.net/blog/fix-for-java-file-encoding-problems-with-gradle.html
-tasks.withType(JavaCompile).configureEach {
-	options.encoding = "UTF-8"
+	// TODO: jar { from("LICENSE") }?
```
Comment (Fallen-Breath) [inline build.gradle]: Try to keep the license file inside the jar

File: build.gradle [hunk 2]
```diff
@@ -1,70 +1,69 @@
...
+		// Add repositories to retrieve artifacts from in here.
+		// You should only use this when depending on other mods because
+		// Loom adds the essential maven repositories to download Minecraft and libraries from automatically.
+		// See https://docs.gradle.org/current/userguide/declaring_repositories.html
+		// for more information about repositories.
+	}
+
+	tasks.withType(JavaCompile) {
+		options.encoding = "UTF-8"
+		options.release = 17
```
Comment (Fallen-Breath) [inline build.gradle]: will this affect java8 compatibility

File: common/build.gradle [hunk 3]
```diff
@@ -0,0 +1,18 @@
+architectury {
+	common(rootProject.enabled_platforms.split(","))
+}
+
+dependencies {
+	// We depend on fabric loader here to have `org.objectweb.asm` and `org.spongepowered.asm`
+	// Do NOT use other classes from fabric loader
```
Comment (Fallen-Breath) [inline common/build.gradle:7]: is it possible to only depend on `org.spongepowered.asm`

File: fabric/build.gradle [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	id "com.github.johnrengelman.shadow" version "7.1.2"
+}
+
+architectury {
+	platformSetupLoomIde()
+	fabric()
+}
+
+loom {
+	accessWidenerPath = project(":common").loom.accessWidenerPath
```
Comment (Fallen-Breath) [inline fabric/build.gradle]: is this useless? conditional-mixin does not and will not need any access wideners

File: fabric/src/main/java/me/fallenbreath/conditionalmixin/api/util/fabric/VersionCheckerImpl.java [hunk 5]
```diff
@@ -1,20 +1,29 @@
...
-public class VersionChecker
-{
-	/**
-	 * Use fabric-loader's util to check if a given version satisfy a version predicate
-	 * @param version A Version object from fabric. You can use {@link Version#parse} to create one
-	 * @param versionPredicate A string indicates a version predicate, e.g. ">=1.2.0" or "2.0.x"
-	 */
-	public static boolean doesVersionSatisfyPredicate(Version version, String versionPredicate)
-	{
+public class VersionCheckerImpl {
```
Comment (Fallen-Breath) [inline fabric/src/main/java/me/fallenbreath/conditionalmixin/api/util/fabric/VersionCheckerImpl.java]: missing newline. keep the styling. same for other VersionCheckerImpl

File: fabric/src/main/java/me/fallenbreath/conditionalmixin/api/util/fabric/VersionCheckerImpl.java [hunk 6]
```diff
@@ -1,20 +1,29 @@
...
+    public static boolean isModPresent(String modId)
+    {
+        return FabricLoader.getInstance().isModLoaded(modId);
+    }
+
+    public static String getVersionString(String modId)
```
Comment (Fallen-Breath) [inline fabric/src/main/java/me/fallenbreath/conditionalmixin/api/util/fabric/VersionCheckerImpl.java]: seems to be a new API. would returning optional / nullable be better than throwing an exception

File: forge/src/main/java/me/fallenbreath/conditionalmixin/forge/ConditionalMixinForge.java [hunk 7]
```diff
@@ -0,0 +1,8 @@
+package me.fallenbreath.conditionalmixin.forge;
+
+import me.fallenbreath.conditionalmixin.ConditionalMixinMod;
+import net.minecraftforge.fml.common.Mod;
+
+@Mod(ConditionalMixinMod.MOD_ID)
+public class ConditionalMixinForge {
```
Comment (Fallen-Breath) [inline forge/src/main/java/me/fallenbreath/conditionalmixin/forge/ConditionalMixinForge.java]: another missing newline xd

File: forge/src/main/resources/META-INF/mods.toml [hunk 8]
```diff
@@ -0,0 +1,18 @@
+modLoader = "javafml"
+loaderVersion = "[1,)"
```
Comment (Fallen-Breath) [inline forge/src/main/resources/META-INF/mods.toml:2]: does this line say that it support forge loader since version 1? is this correct?

File: build.gradle (hunk 1 above)
Comment (RubixDev) [reply build.gradle]: done

File: common/build.gradle (hunk 3 above)
Comment (RubixDev) [reply common/build.gradle:7]: the Architectury template I based this on also depends on fabric loader in the common module, but they do it for the `@Environment` annotation. I only edited the comment there because in this case we need those two packages which are also available in both fabric and forge environments, but we need to tell gradle that so we can use it in the common module. If you really want me to change that, I'll see what I can do

File: fabric/build.gradle (hunk 4 above)
Comment (RubixDev) [reply fabric/build.gradle]: yes, seems like I missed that bit

File: fabric/src/main/java/me/fallenbreath/conditionalmixin/api/util/fabric/VersionCheckerImpl.java (hunk 5 above)
Comment (RubixDev) [reply fabric/src/main/java/me/fallenbreath/conditionalmixin/api/util/fabric/VersionCheckerImpl.java]: sorry, missed that one

File: fabric/src/main/java/me/fallenbreath/conditionalmixin/api/util/fabric/VersionCheckerImpl.java (hunk 6 above)
Comment (RubixDev) [reply fabric/src/main/java/me/fallenbreath/conditionalmixin/api/util/fabric/VersionCheckerImpl.java]: done

File: forge/src/main/java/me/fallenbreath/conditionalmixin/forge/ConditionalMixinForge.java (hunk 7 above)
Comment (RubixDev) [reply forge/src/main/java/me/fallenbreath/conditionalmixin/forge/ConditionalMixinForge.java]: same here, missed it

File: forge/src/main/resources/META-INF/mods.toml (hunk 8 above)
Comment (RubixDev) [reply forge/src/main/resources/META-INF/mods.toml:2]: yes, I temporarily left it at that because unlike fabric, the forge loader changes version with every Minecraft version and the goal is to support as far back as possible. But I will still need to test what is actually the minimum supported version

**Comment by RubixDev** [comment]
> Some questions:
> 
> 1. Any instructions on how you set up this project, like basing on an example mod? Providing the source might help for better maintainability on the setup
> 2. I see gradle stuffs are totally rewritten. It's fine for the maven publish thing, I can fix it latter. Have you test if jitpack works
> 3. By using jitpack, you can test if your PR works as expected in both platforms, and wide enough mc version range (e.g. 1.14.4 fabric with old loader and java8)

1. The setup is based on the [`1.20.1-forge-fabric.zip`](https://github.com/architectury/architectury-templates/releases/download/release_build_1698078656340/1.20.1-forge-fabric.zip) template from https://github.com/architectury/architectury-templates/releases/tag/release_build_1698078656340
2. jitpack seems to have had some issues recently, but after increasing the available memory for Gradle the latest build worked fine: https://jitpack.io/com/github/RubixDev/conditional-mixin/5e00270aea/build.log
3. so far I've only tested all three platforms on different 1.20 versions, but I'm on it to test older versions that use Java 8

---

With this setup, the publications will have following artifcats:

- `me.fallenbreath:conditional-mixin:<version>` - the common module, users can depend on this in platform independent code (probably as `modCompileOnly`)
- `me.fallenbreath:conditional-mixin-fabric:<version>` - the Fabric mod, used just like before, although it now has the additional `-fabric` in the artifact id
- `me.fallenbreath:conditional-mixin-forge:<version>` - a Forge mod, should be used like the Fabric one but for Forge (`include(modImplementation("..."))`)
- `me.fallenbreath:conditional-mixin-neoforge:<version>`- same as Fabric and Forge but for NeoForge

It would technically also be possible to have one singular artifact which includes all these and decides at runtime which one to use, or with another project setup it would also be possible to manually check for the existence of certain classes and call appropriate code, but I don't think that's worth the hassle. The Architectury project setup should work fine

**Comment by Fallen-Breath** [comment]
> `me.fallenbreath:conditional-mixin:<version>` - the common module, users can depend on this in platform independent code

Suggestion: use `me.fallenbreath:conditional-mixin-common:<version>` for this to keep `me.fallenbreath:conditional-mixin` unused. This is more eplicit, and can prevent existing users from upgrading without noticing the syntax change from `conditional-mixin` to `conditional-mixin-fabric`

File: build.gradle (hunk 2 above)
Comment (RubixDev) [reply build.gradle]: turns out, yes it does, but I've been able to resolve that

**Comment by RubixDev** [comment]
> Suggestion: use `me.fallenbreath:conditional-mixin-common:<version>` for this to keep `me.fallenbreath:conditional-mixin` unused. This is more eplicit, and can prevent existing users from upgrading without noticing the syntax change from `conditional-mixin` to `conditional-mixin-fabric`

Good idea! I changed that.

---

Alright so I did as much testing as I could. Fabric works perfectly fine on all versions. I tested NeoForge on 1.20.2 and 1.20.4 (the only two supported versions afaik) without any issues. Now when it comes to Forge, I didn't even manage to get a normal Forge mod for 1.15 with mixins working, so I couldn't really try adding conditional-mixin. That being said, in my Inventorio setup using it on Forge 1.20.1 works flawlessly and it should also work for older versions as it is also compiled for Java 8. So I think overall this is ready to be merged.

A few more thinks to note:
- because the artifact name(s) are no longer the same as the repo name, jitpack users will now have to specify `com.github.Fallen-Breath.conditional-mixin:conditional-mixin-<platform>:<version>` instead of just `com.github.Fallen-Breath:conditional-mixin:<version>`
- NeoForge requires Java 17 so that is overriden in its `build.gradle`
- the mod id differs between Fabric and (Neo)Forge because the latter doesn't allow dashes `-` in mod ids, it is replaced by an underscore `_`

Also for my personal use case I currently don't even need the version predicates, so it would be awesome if you could release it as is for now and we add the translation stuff later.

File: forge/src/main/resources/META-INF/mods.toml (hunk 8 above)
Comment (RubixDev) [reply forge/src/main/resources/META-INF/mods.toml:2]: I've been unable to test which version is exactly the minimum supported, but I don't think it's a huge issue to leave it as is. Even the official Forge MDK (template mod) sets the loader and forge version ranges to `[0,)` by default.

Comment (Fallen-Breath) [reply forge/src/main/resources/META-INF/mods.toml:2]: ok, just leave it here and fix it when someone finds the incompatibility

**Comment by Fallen-Breath** [comment]
Ah the mod id thing is annoying. Maybe at least you can make the fabric version provides `conditional_mixin` as well, so at least it's viable for those who want to use a single mod id

i.e. in `fabric.mod.json`, add:

```json
  "provides": [
    "conditional_mixin"
  ],
```

Also, for this

> because the artifact name(s) are no longer the same as the repo name, jitpack users will now have to specify com.github.Fallen-Breath.conditional-mixin:conditional-mixin-<platform>:<version> instead of just com.github.Fallen-Breath:conditional-mixin:<version>

I think it's just because the Gradle project now contains multiple subprojects with artifacts

**Comment by RubixDev** [comment]
> Ah the mod id thing is annoying. Maybe at least you can make the fabric version provides `conditional_mixin` as well, so at least it's viable for those who want to use a single mod id
> 
> i.e. in `fabric.mod.json`, add:
> 
> ```json
>   "provides": [
>     "conditional_mixin"
>   ],
> ```

Could you maybe do that quickly? I don't have access to a computer anymore today.

**Comment by Fallen-Breath** [comment]
Done the remaining stuffs. Last thing required before merge: testing its functionality in each platform. I don't have good mod development environments for forge / neoforge though

> so far I've only tested all three platforms on different 1.20 versions, but I'm on it to test older versions that use Java 8

You can make a list / table of what you have tested when you finishes it

**Comment by Fallen-Breath** [comment]
Found an unmentioned breaking change: `me.fallenbreath.conditionalmixin.api.util.VersionChecker#doesVersionSatisfyPredicate` now no longer accepts a version, but accepts a mod id instead. This breaks the method signature and its semantic

Suggestion:

1. Rename the new "`doesVersionSatisfyPredicate`" to "`doesModVersionSatisfyPredicate`"
2. Reintroduce the old `doesVersionSatisfyPredicate`, mark it as `@Deprecated`, and only implements it in the fabric platform. You can use `Object` as the paramenter type to avoid depending fabric-loader

**Comment by Fallen-Breath** [comment]
https://github.com/Fallen-Breath/conditional-mixin/pull/5#issuecomment-2016502303 is now resolved. Here's a summary for all changes of `me.fallenbreath.conditionalmixin.api.util.VersionChecker` in this PR

- Added `isModPresent`
- Added `getModVersionString`
- Added `doesModVersionSatisfyPredicate`
- Deprecated `doesVersionSatisfyPredicate`

**Comment by RubixDev** [comment]
> You can make a list / table of what you have tested when you finishes it

I pretty much summarized my testing in https://github.com/Fallen-Breath/conditional-mixin/pull/5#issuecomment-2006570931, but here it is as a table:

|        | Fabric             | NeoForge           | Forge              |
| ------ | ------------------ | ------------------ | ------------------ |
| 1.14.4 | :heavy_check_mark: | N/A                |                    |
| 1.15.2 | :heavy_check_mark: | N/A                |                    |
| 1.16.5 | :heavy_check_mark: | N/A                |                    |
| 1.17.1 | :heavy_check_mark: | N/A                |                    |
| 1.18.2 | :heavy_check_mark: | N/A                |                    |
| 1.19.4 | :heavy_check_mark: | N/A                |                    |
| 1.20.1 | :heavy_check_mark: | N/A                | :heavy_check_mark: |
| 1.20.2 | :heavy_check_mark: | :heavy_check_mark: |                    |
| 1.20.4 | :heavy_check_mark: | :heavy_check_mark: |                    |

If you want me to, I can do more Forge testing and maybe even some Legacy Fabric tomorrow.

**Comment by Fallen-Breath** [comment]
Nice. Have you tested this PR with java8 in old mc versions by the way?

**Comment by RubixDev** [comment]
The Fabric tests before 1.17 were done with Java 8

**Comment by Fallen-Breath** [comment]
Then the last remaing test is on an old forge version (e.g. mc 1.16), to check if the used forged api still works. After that, it's ready to go

**Comment by RubixDev** [comment]
I'll do that tomorrow

**Comment by RubixDev** [comment]
Alright, I've spent too many hours trying to fight various Gradle errors without much luck. Here's the updated table, which I understand is a bit lackluster, but I'm done with Gradle for today :face_exhaling:

|        | Fabric             | NeoForge           | Forge              |
| ------ | ------------------ | ------------------ | ------------------ |
| 1.14.4 | :heavy_check_mark: | N/A                | :question:[^1][^2] |
| 1.15.2 | :heavy_check_mark: | N/A                | :question:[^3]     |
| 1.16.5 | :heavy_check_mark: | N/A                | :heavy_check_mark: |
| 1.17.1 | :heavy_check_mark: | N/A                | :heavy_check_mark: |
| 1.18.2 | :heavy_check_mark: | N/A                | :question:[^4]     |
| 1.19.4 | :heavy_check_mark: | N/A                | :question:[^4]     |
| 1.20.1 | :heavy_check_mark: | N/A                | :heavy_check_mark: |
| 1.20.2 | :heavy_check_mark: | :heavy_check_mark: | :heavy_check_mark: |
| 1.20.4 | :heavy_check_mark: | :heavy_check_mark: | :question:[^5]     |

[^1]: User has to manually depend on Mixin, as it's not bundled in Forge.

[^2]: `java.lang.NoSuchFieldException: ucp`

    I'm having trouble getting Forge to run at all

[^3]: `java.lang.NoClassDefFoundError: jdk/nashorn/api/scripting/NashornScriptEngineFactory`

    I assume this is not related to conditional-mixin, but I just wasn't able to get this running.

[^4]: `java.lang.module.ResolutionException: Module commons.logging reads more than one module named cpw.mods.securejarhandler`

    Again, I don't think this has anything to do with conditional-mixin, but I didn't get past this Gradle error.

[^5]: `java.lang.module.FindException: Module org.lwjgl.openal not found, required by org.lwjgl.openal.natives`

    Another Gradle error...

**Comment by Fallen-Breath** [comment]
Thanks a lot for your testing! I think these tests are enough as 1.16/1.17 forge arr tested

**Review by Fallen-Breath — APPROVED** [review]

**Comment by RubixDev** [comment]
I also just did a quick test on Legacy Fabric for Minecraft 1.7.10 and that also worked.
